### PR TITLE
Add eusbullet for using bullet collision function (implement with c defun)

### PR DESCRIPTION
**Pull request by mmurooka**
reimplement  https://github.com/euslisp/jskeus/pull/538 according to advice of https://github.com/euslisp/jskeus/pull/538#discussion_r300599272.
Defference with previous PR is this PR uses C defun and does not use defforeign.

Correspondence between PQP and new Bullet is:
pqp.l <-> bullet.l
euspqp.c <-> eusbullet.c
CPQP.C <-> CBULLET.cpp
PQP (directory) <-> None (bullet is installed by apt)

### CODE REVIEW

**Comment by k-okada** [comment]
Thank you for contributing jskeus documentation<br>
Please check latest documents before merging<br>

PDF version of Japanese jmanual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/431/artifacts/0//home/circleci/project/artifacts/jmanual.pdf' target='_blank'>jmanual.pdf</a></strong>
HTML version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/431/artifacts/0//home/circleci/project/artifacts/html/jmanual.html' target='_blank'>jmanual.html</a></strong>
Sphinx (ReST) version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/431/artifacts/0//home/circleci/project/artifacts/rst/html/jmanual.html' target='_blank'>jmanual.rst</a></strong>

**Comment by mmurooka** [comment]
New manual of collision function: https://431-17358571-gh.circle-artifacts.com/0/home/circleci/project/artifacts/rst/html/irtcollision.html

File: irteus/irtcollision.l [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+         (apply #'pqp-collision-distance model1 model2 args))
+        ((eq *collision-algorithm* *collision-algorithm-bullet*)
+         (apply #'bt-collision-distance model1 model2 args))
+        (t
+         (error "invalid collision algorithm: ~a~%" *collision-algorithm*)))
+  )
+
+(defun collision-check
+    (model1 model2 &rest args &key &allow-other-keys)
+  "Check collision between model1 and model2 using Bullet.
```
Comment (Naoki-Hiraoka) [inline irteus/irtcollision.l]: I think ```using Bullet``` should be removed. 
https://github.com/mmurooka/jskeus/pull/2

File: doc/irtcollision.tex [hunk 2]
```diff
@@ -1,26 +1,114 @@
...
+
+(dotimes (i 100)
+  (send *sarm* :joint0 :joint-angle 1 :relative t)
+  (send *irtviewer* :draw-objects))
+(send *sarm* :end-coords :dissoc *target*)
+(dotimes (i 100)
+  (send *sarm* :joint0 :joint-angle -1 :relative t)
+  (send *irtviewer* :draw-objects))
+\end{verbatim}
+}
```
Comment (Naoki-Hiraoka) [inline doc/irtcollision.tex]: This demo does not work, and ```pqp-collision-check-objects``` is PQP specific.
Please consider https://github.com/mmurooka/jskeus/pull/3

File: irteus/irtcollision.l (hunk 1 above)
Comment (mmurooka) [reply irteus/irtcollision.l]: Thanks, merged.

File: irteus/test/test-collision.l [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+         (ret)
+         )
+    (send obj1 :set-color (float-vector 1 0 0) 0.5)
+    (send obj2 :set-color (float-vector 0 1 0) 0.4)
+    (objects (list obj1 obj2))
+    (do-until-key
+     ;; move object
+     (incf cnt)
+     (send obj1 :newcoords (make-coords :pos (float-vector (* 500.0 (sin (/ cnt 20.0))) 50 0)))
+     (send *irtviewer* :draw-objects)
```
[the remark this answers is not in the document]
Comment (mmurooka) [reply irteus/test/test-collision.l]: Thanks, merged.

File: doc/irtcollision.tex (hunk 2 above)
Comment (mmurooka) [reply doc/irtcollision.tex]: Thanks, merged. Checked that demo works.

**Comment by mmurooka** [comment]
@Naoki-Hiraoka thank you for your review. I merged your PR and rebased commits.

**Comment by k-okada** [comment]
Thank you for contributing jskeus documentation<br>
Please check latest documents before merging<br>

PDF version of Japanese jmanual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/446/artifacts/0//home/circleci/project/artifacts/jmanual.pdf' target='_blank'>jmanual.pdf</a></strong>
HTML version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/446/artifacts/0//home/circleci/project/artifacts/html/jmanual.html' target='_blank'>jmanual.html</a></strong>
Sphinx (ReST) version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/446/artifacts/0//home/circleci/project/artifacts/rst/html/jmanual.html' target='_blank'>jmanual.rst</a></strong>

File: irteus/bullet.l [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                (length (send b :vertices))
+                ))
+         ))
+  (when margin
+    (btsetmargin m margin))
+  m)
+
+(defmethod cascaded-coords
+  (:make-btmodel
+   (&key (fat 0) vs m)
```
Comment (Naoki-Hiraoka) [inline irteus/bullet.l]: Arguments ```ms``` and ```m``` are local variables.
Please consider https://github.com/mmurooka/jskeus/pull/5

File: irteus/bullet.l [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+(require :irtmath)
+(require :irtgeo)
+
+(in-package "GEOMETRY")
+
+(export '(bt-collision-distance bt-collision-check))
+
+(defun bt-make-model-from-body
+    (b &key (csg (send b :csg)) (margin nil) m)
```
Comment (Naoki-Hiraoka) [inline irteus/bullet.l]: ```m``` is local variable.
Please consider https://github.com/mmurooka/jskeus/pull/6

File: irteus/bullet.l (hunk 4 above)
Comment (mmurooka) [reply irteus/bullet.l]: Thanks, merged.

File: irteus/bullet.l (hunk 5 above)
Comment (mmurooka) [reply irteus/bullet.l]: Thanks, merged.

**Comment by mmurooka** [comment]
@Naoki-Hiraoka thank you again for your review. I merged your PR and rebased commits.

**Comment by k-okada** [comment]
Thank you for contributing jskeus documentation<br>
Please check latest documents before merging<br>

PDF version of Japanese jmanual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/457/artifacts/0//home/circleci/project/artifacts/jmanual.pdf' target='_blank'>jmanual.pdf</a></strong>
HTML version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/457/artifacts/0//home/circleci/project/artifacts/html/jmanual.html' target='_blank'>jmanual.html</a></strong>
Sphinx (ReST) version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/457/artifacts/0//home/circleci/project/artifacts/rst/html/jmanual.html' target='_blank'>jmanual.rst</a></strong>

File: irteus/CBULLET.cpp [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+#endif
+
+
+#if HAVE_BULLET
+struct btDistanceInfo
+{ // this class is copied from https://github.com/bulletphysics/bullet3/blob/master/test/collision/btDistanceInfo.h
+  btVector3 m_pointOnA;
+  btVector3 m_pointOnB;
+  btVector3 m_normalBtoA;
+  btScalar m_distance;
```
Comment (Naoki-Hiraoka) [inline irteus/CBULLET.cpp]: Bullet_User_Manual.pdf によれば，```btScalar```型はデフォルトでは```float```になっており，コンパイル時に```BT_USE_DOUBLE_PRECISION```をONにすると```double```になるそうです．このオプションを付けなくても十分に機能していますので必須ではありませんが，64bitマシンの場合にはこのオプションをつけてコンパイルした方が正確になりそうです．

File: irteus/CBULLET.cpp [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+};
+
+long BT_MakeCapsuleModel(double radius, double height)
+{
+  return (long)(new btCapsuleShapeZ(radius, 0.5*height));
+};
+
+long BT_MakeMeshModel(double *verticesPoints, long numVertices)
+{
+  btConvexHullShape* pshape = new btConvexHullShape();
```
Comment (Naoki-Hiraoka) [inline irteus/CBULLET.cpp]: よく使用されるロボットモデルで簡単に試した限りでは，現在の実装でも安定性・計算時間ともにpqpと同程度には問題無く機能しているので必須ではありませんが，
Bullet_User_Manual.pdfによれば，パフォーマンスや安定性のためには
```
Use less then 100 vertices in a convex mesh
```
とするのが良く，```btShapeHull utility```を用いてconvex hullを単純化するべきだそうです．

**Comment by mmurooka** [comment]
Review approved. @k-okada Please tell if more refine necessary.

File: .travis.sh [hunk 8]
```diff
@@ -51,6 +51,7 @@ for test_l in irteus/test/*.l; do
 
     # osrf/ubuntu_arm64:trusty takes >50 min, skip irteus-demo.l
     [[ "$DOCKER_IMAGE" == *"arm64:trusty"* && $test_l =~ irteus-demo.l ]] && continue;
+    [[ ( "$DOCKER_IMAGE" == *"trusty"* || "$DOCKER_IMAGE" == *"jessie"* ) && $test_l =~ test-collision.l ]] && continue;
```
Comment (k-okada) [inline .travis.sh]: please add comment why we have to skip this test.

Comment (mmurooka) [reply .travis.sh]: added comment: https://github.com/euslisp/jskeus/pull/555/commits/1f4da2f7da05f3106eea44366a1cfc92cd07277f#diff-b4553af39acae935d505c76e139d4aa3R54-R55
(rebase commit into previous)

**Comment by k-okada** [comment]
Thank you for contributing jskeus documentation<br>
Please check latest documents before merging<br>

PDF version of Japanese jmanual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/473/artifacts/0//home/circleci/project/artifacts/jmanual.pdf' target='_blank'>jmanual.pdf</a></strong>
HTML version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/473/artifacts/0//home/circleci/project/artifacts/html/jmanual.html' target='_blank'>jmanual.html</a></strong>
Sphinx (ReST) version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/473/artifacts/0//home/circleci/project/artifacts/rst/html/jmanual.html' target='_blank'>jmanual.rst</a></strong>

**Comment by k-okada** [comment]
Thank you for contributing jskeus documentation<br>
Please check latest documents before merging<br>

PDF version of Japanese jmanual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/493/artifacts/0//home/circleci/project/artifacts/jmanual.pdf' target='_blank'>jmanual.pdf</a></strong>
HTML version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/493/artifacts/0//home/circleci/project/artifacts/html/jmanual.html' target='_blank'>jmanual.html</a></strong>
Sphinx (ReST) version of Japanese manual: <strong><a href='https://circleci.com/gh/euslisp/jskeus/493/artifacts/0//home/circleci/project/artifacts/rst/html/jmanual.html' target='_blank'>jmanual.rst</a></strong>

**Comment by k-okada** [comment]
@mmurooka all test in EusLisp failing, is this related to this change ? https://travis-ci.org/euslisp/EusLisp/builds/604932545

**Comment by mmurooka** [comment]
it seems so... I'll check.

One of the problem is that, in all test cases, jskeus is compiled without bullet.
https://api.travis-ci.org/v3/job/604932547/log.txt
```
jskeus is compiled without bullet, so you can not use function eusinteger_t C_BT_CalcCollisionDistance(eusinteger_t, eusinteger_t, eusfloat_t*, eusfloat_t*, eusfloat_t*, eusfloat_t*, eusfloat_t*, eusfloat_t*, eusfloat_t*, eusfloat_t*). Please install bullet >= 2.83.
...
RESULT: test-collision-sphere-analytical-pqp
  TEST-NUM: 1
    PASSED:   1
    FAILURE:  0
RESULT: test-collision-sphere-analytical-bullet
  TEST-NUM: 1
    PASSED:   0
[31m    FAILURE:  1[0m
RESULT: test-collision-cube-approx-pqp
  TEST-NUM: 1
    PASSED:   1
    FAILURE:  0
RESULT: test-collision-cube-approx-bullet
  TEST-NUM: 1
    PASSED:   0
[31m    FAILURE:  1[0m
RESULT: test-collision-cylinder-approx-pqp
  TEST-NUM: 1
    PASSED:   1
    FAILURE:  0
RESULT: test-collision-cylinder-approx-bullet
  TEST-NUM: 1
    PASSED:   0
[31m    FAILURE:  1[0m
RESULT: test-collision-mesh-approx-pqp
  TEST-NUM: 1
    PASSED:   1
    FAILURE:  0
RESULT: test-collision-mesh-approx-bullet
  TEST-NUM: 1
    PASSED:   0
[31m    FAILURE:  1[0m
```

**Pull request merged**